### PR TITLE
Fixes for userAgent delay.  Retrieved async now.

### PR DESCRIPTION
**Pull request by bsneed**
- Made userAgent retrieval async.
- Updated context test to prime it.
- Moved it to the dynamic section of context collection.
- Fixes issue #229

### CODE REVIEW

**Review by alanjcharles — APPROVED** [review]

File: Sources/Segment/Plugins/Platforms/Vendors/AppleUtils.swift [hunk 1]
```diff
@@ -65,17 +66,15 @@ internal class iOSVendorSystem: VendorSystem {
...
-        if Thread.isMainThread {
-            userAgent = WKWebView().value(forKey: "userAgent") as? String
-        } else {
-            DispatchQueue.main.sync {
-              userAgent = WKWebView().value(forKey: "userAgent") as? String
+        // BKS: It was discovered that on some platforms there can be a delay in retrieval.
+        // It has to be fetched on the main thread, so we've spun it off
+        // async and cache it when it comes back.
+        if Self.asyncUserAgent == nil {
+            DispatchQueue.main.async {
```
Comment (didiergarcia) [inline Sources/Segment/Plugins/Platforms/Vendors/AppleUtils.swift:73]: Hey any chance you should re-check the condition `asyncUserAgent == nil` once you're in there?  This reminds me of the patterns that would require double-check locking, no?

File: Sources/Segment/Plugins/Platforms/Vendors/AppleUtils.swift [hunk 2]
```diff
@@ -238,16 +238,15 @@ internal class MacOSVendorSystem: VendorSystem {
...
+                Self.asyncUserAgent = WKWebView().value(forKey: "userAgent") as? String
```
Comment (didiergarcia) [inline Sources/Segment/Plugins/Platforms/Vendors/AppleUtils.swift:246]: Same question in here: Should we check the `asyncUserAgent == nil` condition here again?

File: Sources/Segment/Plugins/Platforms/Vendors/AppleUtils.swift (hunk 1 above)
Comment (bsneed) [reply Sources/Segment/Plugins/Platforms/Vendors/AppleUtils.swift:73]: I don't mind if it gets set twice.  It's locked via the @Atomic marker on the var so no race condition on double-set.

File: Sources/Segment/Plugins/Platforms/Vendors/AppleUtils.swift (hunk 2 above)
Comment (bsneed) [reply Sources/Segment/Plugins/Platforms/Vendors/AppleUtils.swift:246]: I don't mind if it gets set twice. It's locked via the @atomic marker on the var so no race condition on double-set.

File: Tests/Segment-Tests/Analytics_Tests.swift [hunk 3]
```diff
@@ -123,6 +123,12 @@ final class Analytics_Tests: XCTestCase {
         let outputReader = OutputReaderPlugin()
         analytics.add(plugin: outputReader)
         
+        // prime the pump for userAgent, since it's retrieved async.
+        let vendorSystem = VendorSystem.current
+        while vendorSystem.userAgent == nil {
+            RunLoop.main.run(until: Date.distantPast)
```
Comment (didiergarcia) [inline Tests/Segment-Tests/Analytics_Tests.swift:130]: Is this a tight loop? basically spinning until that variable is set? If so, does a something like a count down latch fit more here?

**Review by didiergarcia — APPROVED** [review]

**Pull request merged**